src/libraries/System.Private.CoreLib/src/System/Reflection/Pointer.cs

```diff
@@ -39,6 +39,18 @@ public static object Box(void* ptr, Type type)
             return ((Pointer)ptr)._ptr;
         }
 
+        public override unsafe bool Equals(object? obj)
+        {
+            if (obj is Pointer pointer)
+            {
+                return _ptr == pointer._ptr;
```

Comment:
Is comparison of the pointer Type intentionally omitted here?

Either way, this case should be covered by the tests.

Reply:
I knew I've forgotten something. There is no way to get pointers to polymorphic types in .net right? So type hierarchy isn't an issue here?

Reply:
There might be some scenarios for wanting to ignore the type. For instance:

```cs
int* pA = 0x12340000;
uint* pB = (uint*)pA;
Console.WriteLine(pA == pB); // prints 'True'
Console.WriteLine(Pointer.Box(pA, typeof(int)) == Pointer.Box(pB, typeof(uint)); // prints ???
```

Reply:
Hmm interesting, .net is different to C++ in that regard. C++ doesn't allow comparison of different pointer types. Good to know.

So it looks like .net doesn't differentiate between pointer types when it comes to equality, i.e. `short*` and `double*` are also compared solely by their address.

Reply:
> There might be some scenarios for wanting to ignore the type.

You can handle all these scenarios today, by unboxing the Pointer using Pointer.Unbox and comparing what you need yourself. 

The behavior change for Equals is convenience, it is not enabling any fundamentally new capability. We can argue which one is the most convenient behavior to make it the default.

> .net is different to C++ in that regard. 

More precisely, C# is different to C++ here.  And C is different from C++ and it does allow comparison of different pointer types. All C#, C and C++ can run on .NET today.

Reply:
> The behavior change for Equals is convenience

And as with every behavior change, it is also a potential breaking change.

Reply:
> and it does allow comparison of different pointer types

well standard C++ [doesn't allow](https://godbolt.org/z/Kb5P83) this, except for objects within the same type hierarchy.

> The behavior change for Equals is convenience

I'd argue that `a.Equals(a)` evaluating to `false` is a bug.

Reply:
I've only added tests now and left the implementation as is. I didn't add type checking because the unboxed pointers aren't type checked either. `operator==` wasn't added because it is not required to fix `a.Equals(a)`.